### ISSUE TITLE
Everything blows up when starting path is destroyed

### ISSUE THREAD

**Issue by mateusdelbianco** [spraints/resqued#35]
I'm restarting resqued using capistrano, and after 5 deploys (because I set keep_releases to 5) it fails miserably. The same thing used to happen with unicorn and puma. Somehow the executable path is pointing to app/releases/1234 instead of app/current.

First I got a Gemfile not found, that I fixed pointing it to current in my upstart script:

http://blog.willj.net/2011/08/02/fixing-the-gemfile-not-found-bundlergemfilenotfound-error/

https://github.com/puma/puma/issues/329

Probably the master is still running based on the old deleted path, I'm stuck with:
```
app/shared/bundle/ruby/2.1.0/gems/resqued-0.7.12/lib/resqued/config.rb:17:in `read':
  No such file or directory @ rb_sysopen - config/resqued.rb (Errno::ENOENT)
```
Did you guys run into that problem when deploying to production? If yes, can you share the solution?

**Comment by spraints** [comment]
Can you share your upstart script and the step from the capfile that restarts resqued?

**Comment by spraints** [comment]
I thought about this some more, and I think I've got a solution.

> Did you guys run into that problem when deploying to production? If yes, can you share the solution?

We don't, and there are two things we do to avoid the same problem.

First, we don't do atomic deploys. I don't think that's actually the problem, but it's a factor in what you're seeing.

Second, we generate binstubs during the deploy, and exec the resqued binstub without `bundle exec`. I think this is the fix.

I would guess that your upstart script does something like this: `cd app/current && bundle exec resqued config/resqued.rb`. This means that, by the time [resqued captures `$0`](https://github.com/spraints/resqued/blob/765270ac7e7e78091b1d112ffa36f01e6bd73f8e/exe/resqued#L64), it's been translated into something like `app/shared/bundle/ruby/2.1.0/gems/resqued-0.7.12/exe/resqued`. I assume you're restarting resqued with `SIGHUP`. When you do this, [resqued re-execs the original command](https://github.com/spraints/resqued/blob/765270ac7e7e78091b1d112ffa36f01e6bd73f8e/lib/resqued/listener.rb#L39), which is in `app/shared/...`. The bottom line is, `bundle exec` will change the command from `resqued` into something much more specific to the exact version of your app that's running.

I think the solution is to generate a binstub during deployment (e.g. `bundle install --binstubs bin` or `bundle install && bundle binstubs resqued`) and exec that directly (`exec bin/resqued`) from upstart.

**Comment by mateusdelbianco** [comment]
You are right, this is the relevant snippet in my upstart script:

```
kill signal QUIT

chdir /opt/polisat/dragas/current

exec /home/ubuntu/.rbenv/bin/rbenv exec bundle exec resqued config/resqued.rb -p /opt/polisat/dragas/shared/tmp/pids/resque_workers.pid
```

I'm using the `/sbin/reload` upstart command in capistrano, which by default sends a `HUP` signal to the process.

I manually edited the `exe/resqued` file to print `$0` and `Dir.pwd` right after setting the `START_CTX` variable, and the result was:

```
/opt/polisat/dragas/shared/bundle/ruby/2.1.0/bin/resqued
/opt/polisat/dragas/releases/20150305122800
```

I think the problem here is that it's running inside the `releases/20150305122800` path, not `current`.

I'll change my deploy to generate binstubs, remove bundle exec from my upstart script and see if that fixes the problem.

**Comment by spraints** [comment]
> I'll change my deploy to generate binstubs, remove bundle exec from my upstart script and see if that fixes the problem.

Did you have any luck with this?



> I manually edited the `exe/resqued file` to print `$0` and `Dir.pwd` right after setting the `START_CTX` variable, and the result was:
> 
>     /opt/polisat/dragas/shared/bundle/ruby/2.1.0/bin/resqued
>     /opt/polisat/dragas/releases/20150305122800

`$0` looks good, but `Dir.pwd` looks like it'll still cause problems. This probably means that resqued needs to do something akin to [unicorn's pwd detection](https://github.com/defunkt/unicorn/blob/7d10e85c045a80df260dee1a6717483c5037ef56/lib/unicorn/http_server.rb#L55-L63).

**Comment by mateusdelbianco** [comment]
Sorry for not replying earlier.

Using binstubs, I got `bin/resqued` for `$0` and still `releases` for Dir.pwd.

I'm trying #36 right now, and will let you know what happened.

**Comment by mateusdelbianco** [comment]
Fixed in PR #36, using `PWD` environment variable.

**Comment by spraints** [comment]
[v0.7.13](https://rubygems.org/gems/resqued/versions/0.7.13) is released, and [includes the fix](https://github.com/spraints/resqued/blob/master/CHANGES.md#v0713). Thanks again for reporting the problem! :sparkles:
